accepted/2021/preview-features/preview-features.md

```diff
@@ -0,0 +1,418 @@
+blame Visual Studio, without realizing that this is due to a preview feature.
+This can result in negative customer sentiment which we need to avoid.
+
+Also, for Watson reports we likely want to record if preview features were
+turned on. This will allow us to measure how responsible preview features are
+with respect to hangs and crashes.
+
+And lastly this is for preview features that we think are ready for consumption
+and gathering feedback. This is *not* for unfinished feature work. Preview
+features have to be in a state that we'd consider acceptable for a preview
+release. For highly speculative work, we'd likely keep that outside of GA
+builds, e.g. in feature branches or experimental repos, such as `runtimelabs`.
+
+### Can we make breaking changes in preview features?
+
+Yes! That's the whole point of marking features as preview in the first place.
+Please note that breaking change also includes yanking the preview feature in
+its entirety. Preview features should be treated like previews of .NET: we don't
+make promises and reserve the right to make changes.
+
+Halfway through a release cycle we might also decide that certain features won't
+make the cut for RTM quality and we might decide to mark features we shipped in
+a previous preview as `[RequiresPreviewFeatures]`. The reverse is also possible
+(and argubly not a breaking change) but probably less likely.
+
+### Why a single on/off switch as opposed to individual feature switches?
+
+I originally designed it as such:
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <ExperimentalFeatures>FancyGC;StaticInterfaceMembers</ExperimentalFeatures>
+  </PropertyGroup>
+
+</Project>
+```
+
+The upside is that one can be explicit about which preview feature one wants to
+turn on (and thus depend on).
+
+The downsides were:
+
+1. It's a more complicated concept for the user to understand and very different
+   from how previews of the .NET platform itself work. This might increase the
+   barrier of entry, which we don't want.
+
+2. The test matrix for the product gets much more complicated as the combinations
+   of on/states explode quickly.
+
+3. Not all preview features have a convenient entry point (language syntax,
+   API). Having a global preview mode increases airtime for features that
+   otherwise wouldn't have been turned on.
+
+4. We're unlikely to ship a grab bag of preview features and instead are more
+   likely to ship a cohesive feature set that in its totality is still in
+   preview, such as static interface members, which includes runtime work,
+   language work, and compiler work. Said differently, having a single preview
+   mode forces us to jointly coordinate preview features across all layers.
+
+### Why are we recording whether an assembly used preview features?
+
+For two reasons:
+
+1. It can be used for auditing
+2. It is used to enforce transitivity. An assembly that doesn't use preview
```

Comment:
> we only have to look at the assembly level attribute.

I don't follow this -- we allow the attribute to be on members, without being on the assembly. It seems to me that the compiler needs to recognize this attribute throughout compilation.